File: backend-node/src/storage/FileQueueStorage.ts
import fs from 'node:fs';
import path from 'node:path';

import { Mutex } from 'async-mutex';
import { v4 as uuid } from 'uuid';

import { QueueItem } from '../models/QueueItem';
import { Track } from '../models/Track';
import { QueueStorage } from './QueueStorage';

export class FileQueueStorage implements QueueStorage {
  private readonly mutex = new Mutex();
  private readonly queue: QueueItem[] = [];
  private readonly storagePath: string;

  constructor(storagePath: string) {
    this.storagePath = path.resolve(storagePath);
    this.bootstrap();
  }

  private bootstrap(): void {
    if (!fs.existsSync(this.storagePath)) {
      return;
    }
    try {
      const content = fs.readFileSync(this.storagePath, 'utf-8');
      if (content.trim()) {
        const items = JSON.parse(content) as QueueItem[];
        this.queue.push(...items);
      }
    } catch (error) {
      console.error('Failed to initialize queue storage', error);
    }
  }

  async addToQueue(track: Track, addedBy: string): Promise<QueueItem> {
    return this.mutex.runExclusive(async () => {
      const queueItem: QueueItem = {
        id: uuid(),
        track,
        addedBy,
        addedAt: Date.now(),
        position: this.queue.length,
      };
      this.queue.push(queueItem);
      await this.persist();
      return queueItem;
    });
  }

  async removeFromQueue(queueItemId: string): Promise<boolean> {
    return this.mutex.runExclusive(async () => {
      const index = this.queue.findIndex((item) => item.id === queueItemId);
      if (index === -1) {
        return false;
      }
      this.queue.splice(index, 1);
      this.reassignPositions();
      await this.persist();
      return true;
    });
  }

  async getQueue(): Promise<QueueItem[]> {
    return this.mutex.runExclusive(async () => [...this.queue]);
  }

  async getNext(): Promise<QueueItem | null> {
    return this.mutex.runExclusive(async () => this.queue[0] ?? null);
  }

  async popNext(): Promise<QueueItem | null> {
    return this.mutex.runExclusive(async () => {
      if (!this.queue.length) {
        return null;
      }
      const item = this.queue.shift() ?? null;
      this.reassignPositions();
      await this.persist();
      return item;
    });
  }

  async clearQueue(): Promise<void> {
    return this.mutex.runExclusive(async () => {
      this.queue.length = 0;
      await this.persist();
    });
  }

  async reorderQueue(queueItemId: string, newPosition: number): Promise<boolean> {
    return this.mutex.runExclusive(async () => {
      if (newPosition < 0 || newPosition >= this.queue.length) {
        return false;
      }
      const currentIndex = this.queue.findIndex((item) => item.id === queueItemId);
      if (currentIndex === -1) {
        return false;
      }
      const [item] = this.queue.splice(currentIndex, 1);
      this.queue.splice(newPosition, 0, item);
      this.reassignPositions();
      await this.persist();
      return true;
    });
  }

  private reassignPositions(): void {
    this.queue.forEach((item, index) => {
      item.position = index;
    });
  }

  private async persist(): Promise<void> {
    try {
      await fs.promises.mkdir(path.dirname(this.storagePath), { recursive: true });
      await fs.promises.writeFile(this.storagePath, JSON.stringify(this.queue, null, 2), 'utf-8');
    } catch (error) {
      console.error('Failed to persist queue storage', error);
    }
  }
}

File: backend/src/main/kotlin/com/musicstreaming/storage/FileQueueStorage.kt
package com.musicstreaming.storage

import com.musicstreaming.models.QueueItem
import com.musicstreaming.models.Track
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.io.File
import java.util.UUID

/**
 * File-based implementation of QueueStorage
 * Persists queue to a JSON file with async operations
 */
class FileQueueStorage(
    private val storageFile: File = File("queue_storage.json")
) : QueueStorage {
    
    private val json = Json {
        prettyPrint = true
        ignoreUnknownKeys = true
    }
    
    private val mutex = Mutex()
    private val queue = mutableListOf<QueueItem>()
    
    init {
        // Load existing queue on initialization
        if (storageFile.exists()) {
            try {
                val content = storageFile.readText()
                if (content.isNotBlank()) {
                    val items = json.decodeFromString<List<QueueItem>>(content)
                    queue.addAll(items)
                }
            } catch (e: Exception) {
                // Handle persistence errors gracefully - log but don't fail
                System.err.println("Failed to load queue from storage: ${e.message}")
            }
        }
    }

    override suspend fun addToQueue(track: Track, addedBy: String): QueueItem = mutex.withLock {
        val queueItem = QueueItem(
            id = UUID.randomUUID().toString(),
            track = track,
            addedBy = addedBy,
            addedAt = System.currentTimeMillis(),
            position = queue.size
        )
        
        queue.add(queueItem)
        persistQueue()
        
        return queueItem
    }

    override suspend fun removeFromQueue(queueItemId: String): Boolean = mutex.withLock {
        val removed = queue.removeIf { it.id == queueItemId }
        
        if (removed) {
            // Update positions
            queue.forEachIndexed { index, item ->
                queue[index] = item.copy(position = index)
            }
            persistQueue()
        }
        
        return removed
    }

    override suspend fun getQueue(): List<QueueItem> = mutex.withLock {
        return queue.toList()
    }

    override suspend fun getNext(): QueueItem? = mutex.withLock {
        return queue.firstOrNull()
    }

    override suspend fun popNext(): QueueItem? = mutex.withLock {
        if (queue.isEmpty()) {
            return null
        }
        
        val next = queue.removeAt(0)
        
        // Update positions
        queue.forEachIndexed { index, item ->
            queue[index] = item.copy(position = index)
        }
        
        persistQueue()
        
        return next
    }

    override suspend fun clearQueue() = mutex.withLock {
        queue.clear()
        persistQueue()
    }

    override suspend fun reorderQueue(queueItemId: String, newPosition: Int): Boolean = mutex.withLock {
        val currentIndex = queue.indexOfFirst { it.id == queueItemId }
        
        if (currentIndex == -1 || newPosition < 0 || newPosition >= queue.size) {
            return false
        }
        
        val item = queue.removeAt(currentIndex)
        queue.add(newPosition, item)
        
        // Update positions
        queue.forEachIndexed { index, queueItem ->
            queue[index] = queueItem.copy(position = index)
        }
        
        persistQueue()
        
        return true
    }

    /**
     * Persist the queue to file asynchronously
     * Handles errors gracefully without throwing exceptions
     */
    private suspend fun persistQueue() {
        try {
            withContext(Dispatchers.IO) {
                val jsonContent = json.encodeToString(queue)
                storageFile.writeText(jsonContent)
            }
        } catch (e: Exception) {
            // Handle persistence errors gracefully - log but don't fail
            System.err.println("Failed to persist queue to storage: ${e.message}")
        }
    }

    /**
     * Restore queue from file
     * @return List of queue items or empty list if restore fails
     */
    suspend fun restore(): List<QueueItem> = mutex.withLock {
        return try {
            if (storageFile.exists()) {
                withContext(Dispatchers.IO) {
                    val content = storageFile.readText()
                    if (content.isNotBlank()) {
                        json.decodeFromString<List<QueueItem>>(content)
                    } else {
                        emptyList()
                    }
                }
            } else {
                emptyList()
            }
        } catch (e: Exception) {
            // Handle restore errors gracefully
            System.err.println("Failed to restore queue from storage: ${e.message}")
            emptyList()
        }
    }
}
